src/EntityId/PrefixingEntityIdParser.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Wikibase\DataModel\Services\EntityId;
+
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\EntityIdParser;
+
+/**
+ * EntityIdParser that adds a fixed prefix and parses resulting string as an EntityId.
+ * This can be used to prefix IDs of entities coming from a foreign repository
+ * with the repository name to avoid clashes with IDs of local entities.
+ *
+ * @since 3.7
+ *
+ * @license GPL-2.0+
+ */
+class PrefixingEntityIdParser implements EntityIdParser {
+
+	/**
+	 * @var string
+	 */
+	private $prefix;
+
+	/**
+	 * @var EntityIdParser
+	 */
+	private $idParser;
+
+	/**
+	 * @param string $prefix Prefix to be added
+	 * @param EntityIdParser $idParser
+	 */
+	public function __construct( $prefix, EntityIdParser $idParser ) {
+		$this->prefix = $prefix;
+		$this->idParser = $idParser;
+	}
+
+	/**
+	 * Adds the prefix to the given string and parses the resulting string as an EntityId.
+	 * If the string is already prefixed, the prefix is not duplicated (this check is case sensitive).
+	 *
+	 * @param string $idSerialization
+	 * @return EntityId
+	 */
+	public function parse($idSerialization) {
+		// Do not add prefix if $idSerialization already starts with it
+		if ( strncmp( $idSerialization, $this->prefix, strlen( $this->prefix ) ) !== 0 ) {
+			$idSerialization = $this->prefix . $idSerialization;
```

Comment:
I was about to comment on this too. Reading the description again, it seems like the intent here is to be a generic prefixing parser that does not know about the meaning of different parts of IDs. If that where not the case, the class name should also change from something technical to using the correct domain term. As long as it's either way, which it currently is, it's fine by me.

In case we'd want a service specific to foreign repository "prefixes" [0], I'd disagree with your suggestion to use `EntityId::joinSerialization`. Making a static call to code in another component instead of a much simpler single line of code that just does two string concats is a very poor tradeoff IMO. I'd be different if the format was likely to change.

[0] In the PR to EntityId I can't find how these things are called. "Prefixes" is an accurate technical term, though perhaps not the best choice of naming for something with more specific meaning in our domain.